Allow dynamic KC config updates

neil-hickey opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

If this is your first time, please read our contributor guidelines: https://github.com/vmware-tanzu/carvel-kapp-controller/blob/develop/CONTRIBUTING.md and developer guide https://github.com/vmware-tanzu/carvel-kapp-controller/blob/develop/docs/dev.md
-->

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:
<!--
If no issue exists for this change, please create an issue and link it here.
-->
Fixes #179

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. 

-->
```release-note

```

#### Additional Notes for your reviewer:

See [this comment](https://github.com/vmware-tanzu/carvel-kapp-controller/issues/179#issuecomment-1078230688) for some concerns and questions before I implemented this feature

- I had a really hard time trying to test this. The reconciler is fairly dumb, so that felt unnecessary. And the config change I wanted an e2e test, but `kubectl exec` was failing from the test suite - always with errors like 

`Running 'kubectl exec -n kapp-controller deployment/kapp-controller -- bash -c ls'...
OCI runtime exec failed: exec failed: container_linux.go:380: starting container process caused: exec: "bash -c ls": executable file not found in $PATH: unknown
`

##### Review Checklist:

- [ ] Follows the [developer guidelines](https://carvel.dev/shared/docs/latest/development_guidelines/)
- [ ] Relevant tests are added or updated
- [ ] Relevant docs in this repo added or updated
- [ ] Relevant carvel.dev docs added or updated in a separate PR and there's
  a link to that PR
- [ ] Code is at least as readable and maintainable as it was before this
  change

#### Additional documentation e.g., Proposal, usage docs, etc.:

```docs

```

neil-hickey commented:
### Implementation Notes:

#### What happens if I update my ca bundle whilst KC is running?

- I added the PEM for https://self-signed.badssl.com/ to the `caCerts` field of our KC configmap. Verified I got an error with `curl https://self-signed.badssl.com/` before updating the configmap and then after updating I ran `curl https://self-signed.badssl.com/` again whilst `kubectl exec'd` on the container and it worked.

#### But what about KC process? Does it pick up the change?

It does not. I added in a `http.Get("https://self-signed.badssl.com/ "` into one of the controllers. After applying the secret and updating the config - the curl still fails with a bad cert error. Upon running `kubectl rollout restart deployment/kapp-controller -n kapp-controller` that `http.Get()` no longer fails. So any http calls within KC will not get the new certs dynamically until the next restart. This feels like a niche case though, as KC is mostly just a fancy `exec()` for the fetch / template / deploy tools.

> NOTE: If we do ever move to using the go library version of these tools instead of invoking binaries, we will likely need to revisit this implementation

#### What about GoRoutine race conditions whilst `os.Setenv`'ing?

I validated that the `os.Setenv` func and `os.Getenv` are [guarded by a r/w lock](https://github.com/golang/go/blob/master/src/syscall/env_unix.go#L115-L116) in golang. So only one goroutine can access it at once. I also printed out the environment variables in each controller to validate that they were picking up the new values dynamically.

#### Is `os.Getenv()` global / shared amongst go routines?

Yep - checkout my playground example https://go.dev/play/p/mJoS_uwD6yC

#### Do we need to have a `lock` around the `/etc/pki/tls/certs/ca-bundle.crt` file?

Probably not. Again golang has a `lock` around the [write function](https://github.com/golang/go/blob/master/src/internal/poll/fd_unix.go#L369-L402). So we should be good here also (given we are not expecting any other process to update the system file.) - which is the assumption the previous code here held

neil-hickey committed:
Add new watcher for config changes

- Dynamically update config based on secret or configmap updates for KC

Signed-off-by: Neil Hickey <nhickey@vmware.com>

neil-hickey commented:
Hmm, I need to give some more thought on our `append to /etc/pki/tls/certs/ca-bundle.crt` approach. As every update of the config appends the certs to that file. So we get lots of duplication.

neil-hickey committed:
Update predicate for all event types

Signed-off-by: Neil Hickey <nhickey@vmware.com>

neil-hickey commented on Dockerfile:
```diff
@@ -87,6 +87,9 @@ COPY --from=0 /usr/local/bin/kapp .
 COPY --from=0 /go/src/github.com/vmware-tanzu/carvel-kapp-controller/controller kapp-controller
 
+# Copy the ca-bundle so we have an original
+RUN cp /etc/pki/tls/certs/ca-bundle.crt /etc/pki/tls/certs/ca-bundle.crt.orig
```
I chose `RUN cp` over `COPY` because `COPY` was failing with a "file not found error". Seems like something to do with accessing the `/etc/` folder via the `COPY` syntax.

benmoss replied:
I think COPY is for copying from either another build stage or from the build context

benmoss commented:
This seems like a bug:
- Create a config (I just copy/pasted the one from https://carvel.dev/kapp-controller/docs/v0.34.0/controller-config/#controller-configuration-spec)
- See that it gets applied
- Delete that secret
- Configuration doesn't change back

I think it's because we noop if we don't find a ConfigMap or Secret, when instead we need to somehow revert to defaults

benmoss commented:
This workflow seems problematic:
- Create a config (I just copy/pasted the one from https://carvel.dev/kapp-controller/docs/v0.34.0/controller-config/#controller-configuration-spec)
- See that it gets applied
- Delete that secret
- Configuration doesn't change back

neil-hickey commented:
Good find! Yeh, I wasn't really sure about that case. Like if you delete the config accidentally and then everything breaks (proxy config is revert to defaults etc), is that the UX we want?

Seems like you were expecting yes, I wasn't really so sure.

neil-hickey committed:
Remove `populated` bool

- We don't need to check if the secret or config exists anymore - we
  just always set the values (even if empty)
- This allows us to restore the values to defaults when the config is
  deleted

Signed-off-by: Neil Hickey <nhickey@vmware.com>

neil-hickey commented:
> This seems like a bug:
> 
>     * Create a config (I just copy/pasted the one from https://carvel.dev/kapp-controller/docs/v0.34.0/controller-config/#controller-configuration-spec)
> 
>     * See that it gets applied
> 
>     * Delete that secret
> 
>     * Configuration doesn't change back
> 
> 
> I think it's because we noop if we don't find a ConfigMap or Secret, when instead we need to somehow revert to defaults

Good point @benmoss  ! I updated the code a bit to accommodate that case. Basically always apply the config (even if there doesn't exist one) -- therefore just set the default (empty string) values.

benmoss commented:
Nice, I was confused by that `populated` bool anyway, not needing it makes the code easier to understand

neil-hickey committed:
Add skippable flag for addTrustedCerts()

- Allows test to skip the os.File() calls in addTrustedCerts()

Signed-off-by: Neil Hickey <nhickey@vmware.com>

benmoss commented on pkg/config/config.go:
```diff
@@ -45,7 +46,7 @@ type Config struct {
 	noProxy       string
 	skipTLSVerify string
-	populated     bool
+	SkipCertCheck bool
```
Not a big fan of this approach, I would just make it configurable where the certs get read/written to and then have a unit test that uses a temp dir or a Reader/Writer

neil-hickey committed:
Make config paths configurable

- This allows testing of each use case (create, update, delete)
- Added a stub for the other tests to mock the os.File() calls with temp
  files

Signed-off-by: Neil Hickey <nhickey@vmware.com>

cppforlife commented on cmd/controller/run.go:
```diff
@@ -114,6 +114,26 @@ func Run(opts Options, runLog logr.Logger) error {
+		reconciler := kcconfig.NewReconciler(coreClient, runLog.WithName("config"))
+
+		ctrl, err := controller.New("Config", mgr, controller.Options{
+			Reconciler: NewUniqueReconciler(&ErrReconciler{
```
i dont think you want to be using `NewUniqueReconciler(&ErrReconciler{`, instead i think it should just be `reconciler`.

neil-hickey replied:
I opted for this because it handles alot of the `synchronization` problems that could arise by processing the requests one at a time. Would changing it back to the plain reconciler not introduce the potential for race conditions on requests in parallel?

cppforlife replied:
MaxConcurrentReconciles=1 should take care of this.

cppforlife commented on pkg/config/config.go:
```diff
@@ -150,12 +145,36 @@ func (gc *Config) ShouldSkipTLSForAuthority(candidateAuthority string) bool {
+	}
+	defer src.Close()
+
+	dst, err := os.OpenFile(systemCertsFilePath, os.O_CREATE|os.O_RDWR|os.O_TRUNC, os.ModeExclusive)
```
is there more details on what exactly ModeExclusive does?

neil-hickey replied:
I think this should just be `0600` instead of the exclusiveMode.

cppforlife commented on pkg/config/config.go:
```diff
@@ -150,12 +145,36 @@ func (gc *Config) ShouldSkipTLSForAuthority(candidateAuthority string) bool {
+	if err != nil {
+		return fmt.Errorf("Opening certs file: %s", err)
+	}
+	defer dst.Close()
```
we should be checking on close errors. flushing might happen in close for writes.

cppforlife commented on pkg/config/config.go:
```diff
@@ -150,12 +145,36 @@ func (gc *Config) ShouldSkipTLSForAuthority(candidateAuthority string) bool {
+
+	if gc.BackupCaBundlePath != "" && gc.SystemCaBundlePath != "" {
+		backupCertsFilePath = gc.BackupCaBundlePath
+		systemCertsFilePath = gc.SystemCaBundlePath
+	}
+
+	src, err := os.OpenFile(backupCertsFilePath, os.O_RDONLY, os.ModeExclusive)
+	if err != nil {
+		return fmt.Errorf("Opening original certs file: %s", err)
+	}
+	defer src.Close()
+
+	dst, err := os.OpenFile(systemCertsFilePath, os.O_CREATE|os.O_RDWR|os.O_TRUNC, os.ModeExclusive)
+	if err != nil {
+		return fmt.Errorf("Opening certs file: %s", err)
+	}
+	defer dst.Close()
+
+	_, err = io.Copy(dst, src)
+	if err != nil {
+		return fmt.Errorf("Copying certs file: %s", err)
+	}
+
 	if certChain == "" {
 		return nil
 	}
 
-	var file *os.File
-	file, err = os.OpenFile(systemCertsFile, os.O_CREATE|os.O_APPEND|os.O_WRONLY, os.ModeAppend)
+	file, err := os.OpenFile(systemCertsFilePath, os.O_APPEND|os.O_WRONLY, os.ModeAppend)
```
in general this seems racy to me in combination with above.

cppforlife replied:
(specifically, initial write or append are not atomic to readers of the file so they may actually see corrupt view of the file while we are modifying it)

cppforlife replied:
@neil-hickey let me know if you want to talk about this.

cppforlife commented on pkg/config/reconciler.go:
```diff
@@ -0,0 +1,70 @@
+		return o.GetName() == kcConfigName
+	})
+
+	err := controller.Watch(&source.Kind{Type: &v1.ConfigMap{}}, &handler.EnqueueRequestForObject{}, p)
+	if err != nil {
+		return fmt.Errorf("Watching Configmaps: %s", err)
+	}
+
+	err = controller.Watch(&source.Kind{Type: &v1.Secret{}}, &handler.EnqueueRequestForObject{}, p)
+	if err != nil {
+		return fmt.Errorf("Watching Secrets: %s", err)
+	}
```
are these somehow limited to kapp controller namespace?

neil-hickey replied:
no, they are currently not scoped by namespace at all. The config is searched for based on whatever the kubeconfig on file says for a namespace, but not filtered in the reconciler by namespace.

https://github.com/vmware-tanzu/carvel-kapp-controller/blob/develop/pkg/config/config.go#L78-L80

Given this namespace could be dynamic (I'm thinking `tkg-system` or `kapp-controller`) we have two choices, listen for that config across all namespaces (current option) or load the ns from the kubeconfig and filter on that.

neil-hickey replied:
I changed this so that it follows the same logic as `GetConfig()`, i.e it loads the namespace from the kubeconfig.

cppforlife commented on pkg/config/reconciler.go:
```diff
@@ -0,0 +1,70 @@
+	return nil
+}
+
+// nolint: revive
```
whats this for?

joe-kimmel-vmw replied:
@cppforlife we commented at the same time --   the nolint directive is saying "don't throw me a linter error for not having some terrible comment like `// Reconcile is the reconciler method for reconcilers`

cppforlife commented on pkg/config/config.go:
```diff
@@ -170,21 +189,15 @@ func (gc *Config) addTrustedCerts(certChain string) (err error) {
 }
 
 func (gc *Config) configureProxies() {
-	if gc.httpProxy != "" {
```
whats the thinking behind removing empty check? if it's to unset, shouldnt we be using Unsetenv?

neil-hickey replied:
The thinking behind it is: We want to always just set the variable to whatever we are given. So creates, updates and deletes are all just a simple `os.Set()`

If they delete the config, this will be set to an empty string. 

We *could* check if its empty and do the `os.Unset()` but that felt like more code for the same outcome ("" and unset are equivalent)

cppforlife replied:
i dont think we have assurance from various tools that http_proxy= (empty) is equivalent to no http_proxy at all.

neil-hickey replied:
hmm, fair enough. Perhaps that assumption was misinformed. It's probably safer to unset the variable if there is a doubt on the behaviour.

cppforlife commented on pkg/config/config.go:
```diff
@@ -99,10 +102,6 @@ func GetConfig(client kubernetes.Interface) (*Config, error) {
 }
 
 func (gc *Config) Apply() error {
-	if !gc.populated {
```
im curious what was trying to protect before?

neil-hickey replied:
It was a short-circuit. If the config that came back from the cluster was empty then we skip all of the file / system operations and just return early. I think it was an optimization rather than protecting anything in particular.

In our new case, an empty config coming back has meaning (we want to restore defaults) so this bool makes less sense now.

joe-kimmel-vmw commented on pkg/config/reconciler.go:
```diff
@@ -0,0 +1,70 @@
+	return nil
+}
+
+// nolint: revive
```
fwiw as a person who argued in favor of setting the linter to require comments, i fully support nolint directives like this one :)

cppforlife replied:
i do not. i think we should either have linter or not have linter. sprinkling dont lint me comments feels dirty.

neil-hickey committed:
Use pure reconciler instead of NewUniqueReconciler

- explicitly unset environment variables on empty values provided
- error check on close() files

Signed-off-by: Neil Hickey <nhickey@vmware.com>

neil-hickey committed:
Scope config to namespace in kubeconfig

Signed-off-by: Neil Hickey <nhickey@vmware.com>

cppforlife commented on cmd/controller/run.go:
```diff
@@ -114,6 +114,23 @@ func Run(opts Options, runLog logr.Logger) error {
+	{ // add controller for config
+		reconciler := kcconfig.NewReconciler(coreClient, runLog.WithName("config"))
+
+		ctrl, err := controller.New("Config", mgr, controller.Options{
```
nit: Config => config

cppforlife commented on pkg/config/reconciler.go:
```diff
@@ -0,0 +1,78 @@
+var _ reconcile.Reconciler = &Reconciler{}
+
+// AttachWatches configures watches needed for reconciler to reconcile the kapp-controller Config.
+func (r *Reconciler) AttachWatches(controller controller.Controller) error {
```
instead of loading kubeConfig here again, may be we should just pass namespace into this method from outside?

neil-hickey replied:
Hmm, the namespace isn't available outside the method. `Options` struct contains the optional ns to watch, but this could be `""` which would be weird and also possible different from `GetConfig()` which uses the kubeconfig on disk to get the NS.

I guess to guarantee it's the same, I was opting for loading it again in `AttachWatches`. I'm hesitant to change the `GetConfig()` signature to return the ns it's checking in as it's used in a few places.

cppforlife replied:
how about using https://github.com/vmware-tanzu/carvel-kapp-controller/blob/2833f23be56ee9c028d5638bca94bf06f5aabec3/pkg/apiserver/apiserver.go#L44. seems like it's set in kc deployment "dynamically".

cppforlife commented on pkg/config/reconciler.go:
```diff
@@ -0,0 +1,78 @@
+
+	kcConfig, err := GetConfig(r.coreClient)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("getting kapp-controller config: %s", err)
```
if Reconcile returns an error, next reconcile is immediately scheduled. i assume we do not want to immediately schedule another one.

neil-hickey replied:
Wow, deep in the reconciler it says:

```
// The Controller will requeue the Request to be processed again if an error is non-nil or
// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
```

The is somewhat contradictory (though technically not) to the `reconcile.Result{}` struct which indicated that you need to set the bool to true to get a new reconcile.

```
type Result struct {
    // Requeue tells the Controller to requeue the reconcile key.  Defaults to false.
    Requeue bool
```

I suspect you hit this before, I didn't even think to look that far down the code path to find that hidden comment on the Reconciler interface

neil-hickey committed:
Fix: Config --> config

Signed-off-by: Neil Hickey <nhickey@vmware.com>

neil-hickey committed:
Use tmp file and rename to system ca-bundle

- As rename is an atomic operation, this should prevent any read/write
  issues from other programs (think ytt, imgpkg, git) from reading the
  ca-bundle file whilst something is writing to it

Signed-off-by: Neil Hickey <nhickey@vmware.com>

neil-hickey committed:
Log error instead of return in Config Reconcile()

- Turns out returning an error forces the reconciler to requeue and this
  is not what we want
- Use tempDir() for instead of hardcoding
- Remove tempfile after use

Signed-off-by: Neil Hickey <nhickey@vmware.com>

neil-hickey committed:
Give permission to cert dir and not just cert file

- This is required for an os.Rename()
- Pass the ns into the attachwatches() using KAPPCTRL_SYSTEM_NAMESPACE

Signed-off-by: Neil Hickey <nhickey@vmware.com>

neil-hickey commented on Dockerfile:
```diff
@@ -65,8 +65,9 @@ RUN tdnf install -y git openssh-clients sed
-RUN chmod g+w /etc/pki/tls/certs/ca-bundle.crt
+# Give the root group write access to the openssh's root bundle directory
+# so we can rename the certs file with our dynamic config, and append custom roots at runtime
+RUN chmod g+w /etc/pki/tls/certs
```
This was changed to allow os.Rename()

- `w` permission is needed on the entire directory to create files / rename files within
- This is something we will need to change in TCE's overlay files when we release or else this breaks!

neil-hickey committed:
Fix: Properly close all open file handles

Signed-off-by: Neil Hickey <nhickey@vmware.com>
